BLOOM

younesbelkada opened the pull request:
# What does this PR do?

Adding BLOOM models into transformers library - recreating the PR as the old PR has a bad git commit history
Original PR: #17202

- [x] add a generation test with a small model pushed on the hub
- [x] slow tests needs to be modified accordingly
- [x] add final credits to all reviewers in a final commit

thomwolf committed:
adding template

thomwolf committed:
update model

thomwolf committed:
model update

thomwolf committed:
update conf for debug model

thomwolf committed:
update conversion

thomwolf committed:
update conversion script

thomwolf committed:
update conversion script

thomwolf committed:
fix missing keys check

younesbelkada committed:
add tests to test the tokenizer in the local machine

younesbelkada committed:
Change variable name

younesbelkada committed:
add tests on xnli dataset

younesbelkada committed:
add more description

younesbelkada committed:
add descriptions + clearer code

younesbelkada committed:
clearer code

younesbelkada committed:
adding new tests + skipping few tests because of env problems

younesbelkada committed:
change comment

younesbelkada committed:
add dtype on the configuration

younesbelkada committed:
add test embeddings

younesbelkada committed:
add hardcoded test

younesbelkada committed:
fix dtype issue

younesbelkada committed:
adding torch.float16 to config

younesbelkada committed:
adding more metrics (min, max, mean)

younesbelkada committed:
add sum

younesbelkada committed:
now the test passes with almost equal

younesbelkada committed:
add files for conversion - test passes on cpu  gpu

younesbelkada committed:
add final changes

younesbelkada committed:
cleaning code

younesbelkada committed:
add new args in the docstring

younesbelkada committed:
fix one liner function

younesbelkada committed:
remove macros

younesbelkada committed:
remove forward attention

younesbelkada committed:
clean up init funtion

younesbelkada committed:
add comments on the issue

younesbelkada committed:
rm scale mask softmax

younesbelkada committed:
do make style

younesbelkada committed:
fix dtype in init

younesbelkada committed:
fixing for loop on att probs

younesbelkada committed:
fix style with black

younesbelkada committed:
fix style + doc error

younesbelkada committed:
fix and debug CI errors (docs + style)

younesbelkada committed:
some updates

- change new operations
- finally add scaled softmax
- added new args in the config

younesbelkada committed:
add changes

- save sharded models
- final changes on the modeling script

younesbelkada committed:
make use cache working

younesbelkada committed:
test commit

- added a text to test the commit

Co-authored-by: thomasw21 <24695242+thomasw21@users.noreply.github.com>

younesbelkada committed:
final changes

- attention mask change
- generation works on BS176b

Co-authored-by: thomasw21 <24695242+thomasw21@users.noreply.github.com>

younesbelkada committed:
add changes

- comment on alibi
- add TODO on seq length

younesbelkada committed:
changes - model + conversion

younesbelkada committed:
move to correct dir

younesbelkada committed:
put ,

younesbelkada committed:
fex fixes

younesbelkada committed:
fix tokenizer autodoc

younesbelkada committed:
fix minor CI issues

younesbelkada committed:
fix minor CI issues

younesbelkada committed:
fix minor CI issues

younesbelkada committed:
Merge branch 'main' into bigscience176b

younesbelkada committed:
fix style issue

younesbelkada committed:
fix minor import issues

younesbelkada committed:
fix few issues

younesbelkada committed:
remove def main on the test

younesbelkada committed:
add require torch

younesbelkada committed:
replace decorator with 'with'

younesbelkada committed:
fix style

younesbelkada committed:
change to bloom

younesbelkada committed:
add quick fix tokenizer

younesbelkada committed:
fix tokenizer file

younesbelkada committed:
fix tokenizer

- merge tests
- small fixes

younesbelkada committed:
fix import issue

younesbelkada committed:
add bloom to readme

younesbelkada committed:
fix consistency

younesbelkada committed:
Update docs/source/en/model_doc/bloom.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
Apply suggestions from code review

fix comment issues on file headers

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
fix doc issue

younesbelkada committed:
small fix - modeling test

younesbelkada committed:
some changes

- refactor some code
- taking into account reviews
- more tests should pass
- removed pruning tests

younesbelkada committed:
remove useless division

younesbelkada committed:
more tests should pass

younesbelkada committed:
more tests should pass

younesbelkada committed:
more tests should pass

younesbelkada committed:
let's try this one

-add alibi offset
- remove all permutes to make the grad operations work
- finger crossed

younesbelkada committed:
Merge branch 'main' into bigscience176b

younesbelkada committed:
refactor

- refactor code
- style changes
- add new threshold for test

younesbelkada committed:
major changes

- change BLOOM to Bloom
- add quick doc on bloom.mdx
- move embeddings test on modeling test

younesbelkada committed:
modify readme

younesbelkada committed:
small fixes

younesbelkada committed:
small fix

- better threshold for a test

younesbelkada committed:
remove old test file from fetcher

younesbelkada committed:
fix small typo

younesbelkada committed:
major change

- change BloomLMHead to BloomForCausalLM

younesbelkada committed:
remove onnx config

younesbelkada committed:
major changes

- refactor the code
- remove asserts
- change tol for test

younesbelkada committed:
make style

younesbelkada committed:
small change

younesbelkada committed:
adding a slow test + commenting old ones for now

younesbelkada committed:
make style

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
make style

younesbelkada committed:
fix duplicates

younesbelkada committed:
cleaning comments on config

younesbelkada committed:
clean a bit conversion file

younesbelkada committed:
refacor a bit modeling file

younesbelkada committed:
refactor tokenizer file

younesbelkada committed:
fix tokenization test issue

younesbelkada committed:
fix tokenization issue #2

younesbelkada committed:
fix tokenization issue second try

younesbelkada committed:
Merge branch 'bigscience176b' of https://github.com/younesbelkada/transformers into bigscience176b

younesbelkada committed:
fix test issue

younesbelkada committed:
make style + add suggestions

younesbelkada committed:
change test fetcher

younesbelkada committed:
try this one

- slow tests should pass
- finger crossed

younesbelkada committed:
possible final changes

younesbelkada committed:
make style

younesbelkada committed:
try fix padding side issue

younesbelkada committed:
fix side

younesbelkada committed:
fix padding issue

younesbelkada committed:
fix ko-readme

younesbelkada committed:
fix config auto

younesbelkada committed:
cleaning modeling file

younesbelkada committed:
keep bloom in caps in ko

younesbelkada committed:
update config docs

younesbelkada committed:
remove pretraining_pp

younesbelkada committed:
remove model parallel

younesbelkada committed:
update config

- add correct config files

younesbelkada committed:
fix duplicates

younesbelkada committed:
fix fetcher

younesbelkada committed:
fix refactor issue

- remove divide function

younesbelkada committed:
try to remove alibi

younesbelkada committed:
small fixes

- fix alibi
- remove seq length
- refactor a bit the code

younesbelkada committed:
put correct values

- fix bos and eos token ids

younesbelkada committed:
fix attention mask loop

Co-authored-by: thomasw21 <24695242+thomasw21@users.noreply.github.com>

younesbelkada committed:
small fixes:

- remove skip bias add

younesbelkada committed:
small fixes

- fix typo in readme
- fix typos in config

younesbelkada committed:
small changes

- remove a test
- add reconstruction test
- change config

younesbelkada committed:
small changes

- change Scaled Softmax to BloomScaledSoftmax

younesbelkada committed:
small fixes

- fix alibi dtype

younesbelkada committed:
major changes

- removing explicit dtype when loading modules
- fixing test args (torch_dtype=auto)
- add dosctring

younesbelkada committed:
fix readmes

younesbelkada committed:
major changes

- now bloom supports alibi shifting
- refactor a bit the code
- better test tolerance now

younesbelkada committed:
refactor a bit

younesbelkada committed:
refactor a bit

younesbelkada committed:
put correct name on test

younesbelkada committed:
change docstring

younesbelkada committed:
small changes

- fix docstring modeling
- fix test tolerance

younesbelkada committed:
fix small nit

- take dtype from tensors in the conversion script

younesbelkada committed:
minor fix

- fix mdx issue

younesbelkada committed:
minor fix

- change config docstring

sIncerass committed:
forward contrib credits from PR14084

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

younesbelkada committed:
apply modifications

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

younesbelkada committed:
resolve softmax upcast

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

younesbelkada committed:
Update src/transformers/models/bloom/modeling_bloom.py

Co-authored-by: Niklas Muennighoff <n.muennighoff@gmail.com>

younesbelkada committed:
final changes modeling

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

younesbelkada committed:
Merge commit 'd156898f3b9b2c990e5963f5030a7143d57921a2'

younesbelkada committed:
merge commit

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

younesbelkada committed:
apply suggestions

Apply suggestions from Stas comments
Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

younesbelkada committed:
Merge branch 'main' into bigscience176b

younesbelkada commented:
The commit's history should be better now..! @patrickvonplaten

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17474). All of your documentation changes will be reflected on that endpoint.

stas00 commented:
There is an unresolved discussion in the original PR - can't link to it, but it's the first unresolved one in https://github.com/huggingface/transformers/pull/17202#pullrequestreview-986873875

In the future please let's not rush to open new PRs but fix the existing ones. as it makes incomplete discussions unresolved and also reading the PR's history for why a certain decision was done will be more difficult to do.

younesbelkada committed:
Fix gradient checkpointing

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

younesbelkada commented:
I propose to continue the discussions on #17202 until all the conversations are resolved. There is only one left for now

younesbelkada committed:
add slow but exact

younesbelkada committed:
add accelerate compatibility

Co-authored-by: Nicolas Patry <Narsil@users.noreply.github.com>

younesbelkada committed:
forward contrib credits

Co-authored-by: thomasw21 <thomasw21@users.noreply.github.com>
Co-authored-by: sgugger <sgugger@users.noreply.github.com>
Co-authored-by: patrickvonplaten <patrickvonplaten@users.noreply.github.com>
Co-authored-by: Niklas Muennighoff <n.muennighoff@gmail.com>
Co-authored-by: LysandreJik <LysandreJik@users.noreply.github.com>

patrickvonplaten commented on tests/models/bloom/test_modeling_bloom.py:
```diff
@@ -0,0 +1,722 @@
+    def test_hidden_states_transformers(self):
+        # TODO ifelse device
+        cuda_available = torch.cuda.is_available()
+        device = torch.device("cuda:0" if cuda_available else "cpu")
```
```suggestion
        device = torch_device
```

See other tests how this is done

patrickvonplaten commented on tests/models/bloom/test_modeling_bloom.py:
```diff
@@ -0,0 +1,722 @@
+            output_dict_norm["min"][idx] = embeddings_ln.min(dim=-1).values[0][i].item()
+            output_dict_norm["max"][idx] = embeddings_ln.max(dim=-1).values[0][i].item()
+            output_dict_norm["mean"][idx] = embeddings_ln.mean(dim=-1)[0][i].item()
+        # word_embeddings
```
don't understand this comment

younesbelkada replied:
It's a useless comment, Will delete it!

patrickvonplaten commented on tests/models/bloom/test_modeling_bloom.py:
```diff
@@ -0,0 +1,722 @@
+        for i, key in enumerate(output_dict_norm.keys()):
+            for j, idx in enumerate(output_dict[key].keys()):
+                self.assertAlmostEqual(EMBEDDINGS_DS_AFTER_LN[key][idx], output_dict_norm[key][idx], places=1)
+        # self.assertDictEqual(EMBEDDINGS_DS_AFTER_LN, output_dict_norm)
```
```suggestion
        # self.assertDictEqual(EMBEDDINGS_DS_AFTER_LN, output_dict_norm)
```

delete or not?

patrickvonplaten commented on tests/models/bloom/test_modeling_bloom.py:
```diff
@@ -0,0 +1,722 @@
+        }
+
+        tensor_ids = torch.LongTensor([EXAMPLE_IDS])
+        # position_ids = torch.LongTensor([[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]])
```
delete or not?

younesbelkada replied:
yes !

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,971 @@
+        )
+
+        # change view to [batch_size, num_heads, q_length, k_length]
+
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,971 @@
+        attention_scores = matmul_result.view(*output_size)
+
+        # attention scores and attention mask [b, np, sq, sk]
+
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,971 @@
+
+        # matmul: [batch_size * num_heads, q_length, head_dim]
+        context_layer = torch.bmm(attention_probs_reshaped, value_layer.transpose(0, 1))
+        # context_layer = torch.bmm(attention_probs, value_layer)
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,971 @@
+
+        # change view [k_length, batch_size x num_heads, head_dim]
+        value_layer = value_layer.transpose(1, 0).reshape(value_layer.size(1), output_size[0] * output_size[1], -1)
+        # value_layer = value_layer.contiguous().view(value_layer.size(1), output_size[0] * output_size[1], -1)
```
```suggestion
```

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,971 @@
+        context_layer = context_layer.view(*new_context_layer_shape)
+
+        # =================
+        # Output. [q_length, batch_size, hidden_size]
```
why this special format for the comment?

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,971 @@
+
+    def _init_weights(self, module):
+        """Initialize the weights."""
+        if isinstance(module, (nn.Linear, Conv1D)):
```
Does Bloom have `Conv1D` layers?

patrickvonplaten replied:
Think we can remove this no?

patrickvonplaten reviewed (approved):
Left some more nits - overall this looks good to me

younesbelkada committed:
fix torch device on tests

younesbelkada committed:
make style

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

younesbelkada committed:
fix nits

Co-authored-by: patrickvonplaten<patrickvonplaten@users.noreply.github.com>

younesbelkada committed:
remove final nits

sgugger commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,965 @@
+            scaling factor used in input tensor scaling.
+    """
+
+    def __init__(
+        self,
+        scaled_masked_softmax_fusion,
+        mask_func,
+        softmax_in_fp32,
+        scale,
+    ):
```
```suggestion
    def __init__(self, scaled_masked_softmax_fusion, mask_func, softmax_in_fp32, scale):
```

sgugger commented on tests/models/bloom/test_modeling_bloom.py:
```diff
@@ -0,0 +1,717 @@
+        # append to next input_ids and token_type_ids
+        next_input_ids = torch.cat([input_ids, next_tokens], dim=-1)
+
+        # output_from_no_past = model(next_input_ids, output_hidden_states=True)["hidden_states"][-1]
+        # output_from_past = model(next_tokens, past_key_values=past, output_hidden_states=True)["hidden_states"][-1]
```
To clean up?

younesbelkada replied:
Fixed in 5003f509c85b110ff4d9fd41a3ec1d6113f37e6e

sgugger commented on tests/models/bloom/test_modeling_bloom.py:
```diff
@@ -0,0 +1,717 @@
+            dim=1,
+        )
+
+        # get two different outputs
+        # output_from_no_past = model(next_input_ids, attention_mask=attn_mask, output_hidden_states=True)["hidden_states"][-1]
+        # output_from_past = model(next_tokens, past_key_values=past, attention_mask=attn_mask, output_hidden_states=True)["hidden_states"][-1]
```
Same?

younesbelkada replied:
Fixed in 5003f509c85b110ff4d9fd41a3ec1d6113f37e6e

sgugger reviewed (approved):
A couple more nits :-)
Also could you make sure slow and non-slow tests pass on one GPU and two GPUs before merging?

younesbelkada committed:
fix doc

- add more details on the doc
- add links to checkpoints

younesbelkada commented on docs/source/en/model_doc/bloom.mdx:
```diff
@@ -0,0 +1,47 @@
+- 1.3B parameters 
+- 2.5B parameters
+- 6.3B parameters
+- 175B parameters 
```
Yes, added also the links to the checkpoints in 41d257337ecfa0b6724527014f29b8f87d16e39d

younesbelkada committed:
Update src/transformers/__init__.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
Update src/transformers/models/bloom/modeling_bloom.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
apply suggestions

Co-authored-by: sgugger <sgugger@users.noreply.github.com>

younesbelkada commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,965 @@
+        use_cache=False,
+        output_attentions=False,
+    ):
+        # hidden_states: [b, s, h]
```
Perfect yes, Fixed in 5003f509c85b110ff4d9fd41a3ec1d6113f37e6e

younesbelkada commented on tests/models/bloom/test_modeling_bloom.py:
```diff
@@ -0,0 +1,717 @@
+
+    all_model_classes = (BloomModel, BloomForCausalLM) if is_torch_available() else ()
+    all_generative_model_classes = (BloomForCausalLM,) if is_torch_available() else ()
+    all_parallelizable_model_classes = (BloomForCausalLM,) if is_torch_available() else ()
```
Thanks for the nit! Fixed in 5003f509c85b110ff4d9fd41a3ec1d6113f37e6e

younesbelkada commented on tests/models/bloom/test_modeling_bloom.py:
```diff
@@ -0,0 +1,717 @@
+    all_parallelizable_model_classes = (BloomForCausalLM,) if is_torch_available() else ()
+    fx_compatible = False
+    test_missing_keys = False
+    test_model_parallel = False
```
Fixed and tested it in 5003f509c85b110ff4d9fd41a3ec1d6113f37e6e

younesbelkada commented:
Should have fixed the nits, will quickly test the slow tests now

younesbelkada committed:
put test torchscript to false

stas00 reviewed (approved):
I have only reviewed parts, and trusting that others did the other parts.

Great job, @younesbelkada - thank you for bearing with the process!

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,958 @@
+            output_tensor = torch.zeros_like(context_layer)
+            for i in range(self.pretraining_tp):
+                output_tensor = output_tensor + nn.functional.linear(
+                    context_layer[:, :, int(i * slices) : int((i + 1) * slices)],
```
@younesbelkada @stas00, we just noticed that weights of a linear layer are always simply passed to `nn.functional.linear`? 

Why do we do this here?

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,958 @@
+        if self.pretraining_tp > 1 and self.slow_but_exact:
+            slices = context_layer.shape[-1] / self.pretraining_tp
+            output_tensor = torch.zeros_like(context_layer)
+            for i in range(self.pretraining_tp):
```
Is this only needed for pretraining or also for inference?

younesbelkada replied:
Hmm for now I think we have to keep them

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,958 @@
+        if self.pretraining_tp > 1 and self.slow_but_exact:
+            slices = context_layer.shape[-1] / self.pretraining_tp
+            output_tensor = torch.zeros_like(context_layer)
+            for i in range(self.pretraining_tp):
```
Is this only needed for pretraining or also for inference?

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,958 @@
+            slices = context_layer.shape[-1] / self.pretraining_tp
+            output_tensor = torch.zeros_like(context_layer)
+            for i in range(self.pretraining_tp):
+                output_tensor = output_tensor + nn.functional.linear(
```
Also not a huge fan of this tbh, I'd prefer to have the weight `self.dense` be different if `self.pretraining_tp > 1`

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,958 @@
+    return alibi
+
+
+def pre_process_alibi_for_pad(alibi, attention_mask, num_heads):
```
What is alibi?

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,958 @@
+        # Layer norm at the beginning of the transformer layer.
+        layernorm_output = self.input_layernorm(hidden_states)
+
+        # Compute alibi tensor
```
what is an alibi tensor? 

What is it's use? Would be great if we could document this a bit more actually

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,958 @@
+            intermediate_output = nn.functional.linear(hidden_states, self.dense_4h_to_h.weight)
+
+        output = intermediate_output
+        output_bias = self.dense_4h_to_h.bias
```
I assume we don't do `output + output_bias` here because it's more efficient to do it later all at once? 
Does it make that big of a difference? 
In my opinion, it'd be much more readable if we do the addition here and just pass `hidden_states` as a single tensor (but ok for me to leave as is if it gives a big speed up)

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,958 @@
+                    self.dense.weight[:, int(i * slices) : int((i + 1) * slices)],
+                )
+        else:
+            output_tensor = nn.functional.linear(context_layer, self.dense.weight)
```
Why not use `self.dense(context_layer)`?

younesbelkada replied:
Actually my bad, it seems that we can't because we need to pass the bias term as an argument to the fused operation that comes later. The only hack we can do is to have something like 
`self.dense_4h_to_h.bias = None` before calling the submodule and then `intermediate_output = self.dense_4h_to_h(hidden_states)`

younesbelkada committed:
Update src/transformers/models/bloom/modeling_bloom.py

Co-authored-by: justheuristic <justheuristic@gmail.com>

younesbelkada commented:
Hey @justheuristic !
thanks for the nit! Just tested in and the tests passed - I think that we should add an extra test for this sanity check :-)

younesbelkada committed:
fix alibi

- create alibi only once

younesbelkada commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,958 @@
+        current_sequence_length = hidden_states.shape[1]
+        if layer_past is not None:
+            current_sequence_length += layer_past[0].shape[1]
+        alibi = build_alibi_tensor(current_sequence_length, self.n_head, layernorm_output.dtype)
```
Thanks for the catch! Yes indeed it can be computed only once. I have proposed a fix in 95f8dc77e3e09a13118eaa6ba2813ecc4e2dfd8f

younesbelkada committed:
add small doc

younesbelkada committed:
make quality

thomasw21 commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,958 @@
+
+        if token_type_ids is not None:
+            token_type_embeds = self.word_embeddings(token_type_ids)
+            token_type_embeds = token_type_embeds.transpose(0, 1).contiguous()
```
Token type embedding should not exist. So I'd expect us to yank that part out of the code, ie have no weights and ignore that input if fed by model.

thomasw21 replied:
Passing it through `word_embeddings` looks wrong.

patrickvonplaten replied:
If the model has no token type embeddings it should definitely not be added here - otherwise the state dict will have random weights attached to it + dead code

younesbelkada replied:
There should be no weights associated with token type embeddings in the code - these embeddings are created using the word embeddings `token_type_embeds = self.word_embeddings(token_type_ids)` 
I did not saw this piece of code in Megatron for now, do we use token type embeddings in Megatron @thomasw21 ? If not I will just completely remove this argument in the modeling file

younesbelkada replied:
Removed the token type emb in 9c2f84a70e014ef82ef81efd1b1fa572d80b9734

younesbelkada committed:
replace torch.nn

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,962 @@
+        self.pretraining_tp = config.pretraining_tp
+        self.slow_but_exact = config.slow_but_exact
+        self.activation_func = bias_gelu_impl
+        self.dense_h_to_4h = nn.Linear(hidden_size, 4 * hidden_size)
```
```suggestion
        self.dense_h_to_4h = nn.Linear(hidden_size, 4 * hidden_size, bias=False)
```

patrickvonplaten replied:
Don't instantiate bias here but use this when needed: https://github.com/huggingface/transformers/blob/babeff5524bf3d5d62cfa70e1297158a755b0810/src/transformers/models/roberta/modeling_roberta.py#L1137

younesbelkada replied:
Should be fixed in 1ca7c1ae27c967216b9f9a48e6e8e2e4026b8dc1

younesbelkada committed:
remove token type emb

younesbelkada commented:
Thank you @patrickvonplaten !
I have benchmarked the performance of fused vs unfused version of the `bias_add` function and I am observing the same performance (with a slightly higher speed with unfused operation). So I will remove that, the bias term will not be passed in a weird manner anymore now

Here is the script to reproduce the benchmarking in case you are interested:

```
import torch, timeit
from transformers import BloomForCausalLM

model = BloomForCausalLM.from_pretrained("bigscience/bigscience-small-testing")
model = model.eval()

model_unfused = BloomForCausalLM.from_pretrained("bigscience/bigscience-small-testing", bias_dropout_fusion=False)
model_unfused = model_unfused.eval()

input_ids = torch.LongTensor([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
n_cycles, n_repeats = 100, 100

def model_fused():
    for _ in range(n_cycles): _ = model(input_ids)

def model_unfused_test():
    for _ in range(n_cycles): _ = model_unfused(input_ids)

print(f'model_fused={timeit.Timer("model_fused()", globals=globals()).timeit(number=n_repeats)}')
print(f'model_unfused_test={timeit.Timer("model_unfused_test()", globals=globals()).timeit(number=n_repeats)}')
```

younesbelkada committed:
fix fused op + output bias

stas00 commented:
[...]
> input_ids = torch.LongTensor([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])

Just be careful when you benchmark to use realistic shapes of tensors - when tiny tensors are used it's unlikely you are benchmarking the real situation. As the framework dominates the compute rather than data processing and you're not benchmarking the performance of the gpu ops.

actually, this benchmark is run on cpu, you probably want to switch it to `.cuda`

younesbelkada commented:
I updated the test with the following snippet:
```
import torch, timeit
from transformers import BloomForCausalLM, BloomTokenizerFast

tokenizer = BloomTokenizerFast.from_pretrained("bigscience/tokenizer")

model = BloomForCausalLM.from_pretrained("bigscience/bigscience-small-testing").cuda()
model = model.eval()

model_unfused = BloomForCausalLM.from_pretrained("bigscience/bigscience-small-testing", bias_dropout_fusion=False).cuda()
model_unfused = model_unfused.eval()

batch_size = 16
input_sequences = ["test sequence"*50 for _ in range(batch_size)]

input_ids = tokenizer(input_sequences, return_tensors='pt')["input_ids"]

n_cycles, n_repeats = 100, 100

def model_fused():
    for _ in range(n_cycles): _ = model(input_ids.cuda())
    torch.cuda.synchronize()

def model_unfused_test():
    for _ in range(n_cycles): _ = model_unfused(input_ids.cuda())
    torch.cuda.synchronize()

print(f'model_fused={timeit.Timer("model_fused()", globals=globals()).timeit(number=n_repeats)}')
print(f'model_unfused_test={timeit.Timer("model_unfused_test()", globals=globals()).timeit(number=n_repeats)}')
```
And I got: 
```
model_fused=321.8526800679999
model_unfused_test=322.2609531270002
```
I guess the difference would be much higher if we increase the sequence length/ batch size.

PS: I emulated the test on google colab with transformers built on a separate branch: `pip install git+https://github.com/younesbelkada/transformers.git@501ceed44d26988a17cd884b73eeb573f1f2bea8`

younesbelkada commented:
I will put both since the fused operation is slower on the CPU!

younesbelkada committed:
add fused op

- now can control fused operation from config

stas00 commented:
On A100 I get:

```
model_fused=35.0372433779994
model_unfused_test=35.158229669999855
```

I suppose choose the most sensible default and allow the user to override it.

If I increase the batch size to 128, the fused version reports to be slower 5%. Give it a try.

We are heading towards using automatic optimizers down the road (torchdynamo/aotautograd/nvfuser) where most of these things will be fused/optimized on the fly, so I think all the custom fusions will become redundant - we should start seeing these tools becoming more common towards the 2nd half of this year.

younesbelkada commented:
Great thanks for the information!
In that case let me put the unfused op by default

younesbelkada committed:
remove fused op

younesbelkada committed:
make quality

thomasw21 commented on src/transformers/models/bloom/configuration_bloom.py:
```diff
@@ -0,0 +1,175 @@
+}
+
+
+class BloomConfig(PretrainedConfig):
```
You should clean up un-used arguments: `scale_attn_by_inverse_layer_idx`, `reorder_and_upcast_attn` (I don't know it that's it)

younesbelkada replied:
Ohh yes thanks for the nit! Fixed in 29072744fa936d89b55ea5fddb2a0cf23daec30a

younesbelkada committed:
small changes

- remove unsed args on config
- removed bias gelu file
- make the model torchscriptable
- add torchscript slow tests

younesbelkada commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,972 @@
+
+class BloomBiasGelu(nn.Module):
+    """
+    BloomBiasGelu wrapper function that make use of the simple function on inference mode to make the model jitable and
```
```suggestion
    BloomBiasGelu wrapper function that make use of the simple function on inference mode to make the model torchscriptable and
```

younesbelkada committed:
Update src/transformers/models/bloom/modeling_bloom.py

younesbelkada committed:
fix slow

younesbelkada committed:
make style

younesbelkada committed:
add accelerate support

younesbelkada commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,969 @@
+    def forward(self, hidden_states, residual):
+        hidden_states = self.bias_gelu(
+            nn.functional.linear(hidden_states, self.dense_h_to_4h.weight), self.dense_h_to_4h.bias
+        )
```
Should have addressed a fix in a24ae86920289e1ef3a03f7ba990e79fcc05e903 ! Please let me know if it works

younesbelkada replied:
(And close the comment if it works!)

thomasw21 commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -483,9 +477,7 @@ def __init__(self, config):
 
     def forward(self, hidden_states, residual):
-        hidden_states = self.bias_gelu(
-            nn.functional.linear(hidden_states, self.dense_h_to_4h.weight), self.dense_h_to_4h.bias
```
I think the whole purpose was that the other operation was jittable into a single fused kernel. that changes now no?

younesbelkada replied:
I think that we finally do not want some operations to be jitable (after doing some benchmarks + from @stas00 's comment here : https://github.com/huggingface/transformers/pull/17474#issuecomment-1146169404
Also I had to define our own GeLU module to make the model torchscriptable (see comments on the class definition). Let me know what do you think!

thomasw21 replied:
can you try with sequence length of 2048? and a bias dimension that's friendly to out usecase?

thomasw21 commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -248,11 +242,11 @@ class BloomBiasGelu(nn.Module):
-            return GeLUFunction.apply(x, bias)
+            return GeLUFunction.apply(x)
         else:
-            return bias_gelu(x, bias)
```
If indeed the change is desirable: we should update the naming of this.

stas00 committed:
add bloom to deepspeed tests

haileyschoelkopf commented:
Hi all! We are working with the smaller BLOOM models in the Bigscience Multilingual Modeling WG and would like to use a `BloomForSequenceClassification` and `BloomForTokenClassification` class in our experiments. 

I'm happy to contribute the code for these, but would it be preferred to let this PR be merged first, or for me to open a PR after this is merged?

younesbelkada commented:
Hi!
Very happy to hear that you want to contribute on that! 
On my side I would say that it is preferable to wait until the PR gets merged. Hopefully very soon ;)

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,961 @@
+    return out
+
+
+def bias_gelu(x):
```
there is no bias anymore no? change naming?

younesbelkada replied:
yes

younesbelkada replied:
fixed in fadb36763ef5b123dae7070108a5fde158b95fb6

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,961 @@
+
+def build_alibi_tensor(max_seq_len, n_head, dtype=torch.bfloat16):
+    """
+    Alibi tensor is not causal as the original paper mentions, it relies on a translation invariance of softmax for
```
can we link to the paper here? Still don't really know what alibi is

justheuristic replied:
https://arxiv.org/abs/2108.12409

justheuristic replied:
@patrickvonplaten , i'm Yozh from bigscience, i need someone's help to brainstorm this code with, would you be available for a quick voicechat?
[if zoom is okay, https://us04web.zoom.us/j/6134661532?pwd=YWUxbHM2OEVjVGtmalo1SWx1dzBsUT09 ]

younesbelkada replied:
thanks all!

younesbelkada replied:
fixed in fadb36763ef5b123dae7070108a5fde158b95fb6

sgugger replied:
This was not fixed in the commit mentioned, could you make sure to have that link here?

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,961 @@
+        # Layer-wise attention scaling
+        self.layer_number = max(1, layer_number)
+        coeff = self.layer_number
+        self.norm_factor = math.sqrt(self.head_dim) * coeff
```
```suggestion
        self.norm_factor = math.sqrt(self.head_dim) * max(1, layer_idx)
```

thomasw21 replied:
```suggestion
        self.norm_factor = math.sqrt(self.head_dim) * self.layer_number
```

younesbelkada replied:
I had to modify a bit - fixed in fadb36763ef5b123dae7070108a5fde158b95fb6

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,961 @@
+        # [batch_size, k_length, num_heads, head_dim] -> [k_length, batch_size * num_heads, head_dim]
+        key_layer = key_layer.transpose(1, 0).reshape(output_size[3], output_size[0] * output_size[1], -1)
+
+        # alibi
```
I don't understand this comment

patrickvonplaten replied:
Also why is it called `matmul_result`? The `matmul` happens only later no?

younesbelkada replied:
Yes true

younesbelkada replied:
fixed in 1d63415e099c53b158be9781d797c86f33a72782

patrickvonplaten commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,961 @@
+        # Final Layer Norm
+        self.ln_f = LayerNorm(self.embed_dim, eps=config.layer_norm_epsilon)
+
+        self.gradient_checkpointing = config.gradient_checkpointing
```
```suggestion
```

We don't want to support the config's gradient_checkpointing anymore

patrickvonplaten replied:
cc @sgugger

sgugger replied:
Good catch, those are all left for backward compatibility in GPT-2 and the like, but we shouldn't have them in a new model indeed (they'll all be removed in v5).

There should be a default to `False` to this however, so we shouldn't remove the line entirely but set:
```suggestion
        self.gradient_checkpointing = False
```
or they'll be errors in the forward.

younesbelkada replied:
Ok got it thanks, initially it was set to `False` but from this comment https://github.com/huggingface/transformers/pull/17202#discussion_r883157770 I had to modify it

patrickvonplaten commented on src/transformers/models/bloom/configuration_bloom.py:
```diff
@@ -0,0 +1,158 @@
+            document](https://huggingface.co/docs/transformers/parallelism) to understand more about it. This value is
+            necessary to ensure exact reproducibility of the pretraining results. Please refer to [this
+            issue](https://github.com/pytorch/pytorch/issues/76232)
+        gradient_checkpointing (`bool`, *optional*, defaults to `True`):
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/bloom/configuration_bloom.py:
```diff
@@ -0,0 +1,158 @@
+        attention_softmax_in_fp32=True,
+        pretraining_tp=1,  # TP rank used when training with megatron
+        dtype="bfloat16",
+        gradient_checkpointing=True,
```
```suggestion
```

patrickvonplaten commented on src/transformers/models/bloom/configuration_bloom.py:
```diff
@@ -0,0 +1,158 @@
+        self.bos_token_id = bos_token_id
+        self.eos_token_id = eos_token_id
+        self.dtype = dtype
+        self.gradient_checkpointing = gradient_checkpointing
```
```suggestion
```

Please also make sure to remove `gradient_checkpointing` from all the configs on the Hub of BLOOM :-)

younesbelkada committed:
minor changes

younesbelkada commented on src/transformers/models/bloom/configuration_bloom.py:
```diff
@@ -0,0 +1,158 @@
+            necessary to ensure exact reproducibility of the pretraining results. Please refer to [this
+            issue](https://github.com/pytorch/pytorch/issues/76232)
+        gradient_checkpointing (`bool`, *optional*, defaults to `True`):
+            Whether to use gradient checkpointing.
```
```suggestion
```

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

younesbelkada committed:
minor change

younesbelkada commented:
Can we merge? I think we have pretty much everything now (Accelerate+DeepSpeed compatibility, etc.) ?

younesbelkada committed:
slow tests pass

younesbelkada commented:
@sgugger I can confirm the slow tests pass on a A100 node in Jean Zay. I have emulated the tests by running `export RUN_SLOW=1` before running the testing script

sgugger commented on docs/source/en/model_doc/bloom.mdx:
```diff
@@ -0,0 +1,47 @@
+
+## Overview
+
+Bloom model has been proposed with its various versions through the [BigScience Workshop](https://bigscience.huggingface.co/). BigScience is inspired by other open science initiatives where researchers have pooled their time and resources to collectively achieve a higher impact.
+The architecture of Bloom is essentially similar to GPT3 (auto-regressive model for next token prediction), but has been trained on different 46 languages including code.
```
```suggestion
The BLOOM model has been proposed with its various versions through the [BigScience Workshop](https://bigscience.huggingface.co/). BigScience is inspired by other open science initiatives where researchers have pooled their time and resources to collectively achieve a higher impact.
The architecture of BLOOM is essentially similar to GPT3 (auto-regressive model for next token prediction), but has been trained on different 46 languages including code.
```

sgugger commented on docs/source/en/model_doc/bloom.mdx:
```diff
@@ -0,0 +1,47 @@
+
+Bloom model has been proposed with its various versions through the [BigScience Workshop](https://bigscience.huggingface.co/). BigScience is inspired by other open science initiatives where researchers have pooled their time and resources to collectively achieve a higher impact.
+The architecture of Bloom is essentially similar to GPT3 (auto-regressive model for next token prediction), but has been trained on different 46 languages including code.
+Several smaller versions of the models have been trained on the same dataset. Bloom is available in the following versions:
```
```suggestion
Several smaller versions of the models have been trained on the same dataset. BLOOM is available in the following versions:
```

sgugger commented on src/transformers/models/bloom/configuration_bloom.py:
```diff
@@ -0,0 +1,154 @@
+        dtype (`str`, *optional*, defaults to `"bfloat16"`):
+            Precision that has been used for the model's training in Megatron. Please load the model in the correct
+            precision by doing `model = BloomModel.from_pretrained(model_name, torch_dtype="auto")`.`
+        pretraining_tp (`int`, *optional*, defaults to `1`):
+            Tensor parallelism rank used during pretraining with Megatron. Please refer to [this
+            document](https://huggingface.co/docs/transformers/parallelism) to understand more about it. This value is
+            necessary to ensure exact reproducibility of the pretraining results. Please refer to [this
+            issue](https://github.com/pytorch/pytorch/issues/76232)
```
I think now this is only used when `slow_but_exact=True`, no? If so, please add this in the docstring :-)

younesbelkada replied:
Yes you are right, should be added in 3a59eb3ffce941403ff5145decb5214978edd390

sgugger commented on src/transformers/models/bloom/modeling_bloom.py:
```diff
@@ -0,0 +1,960 @@
+def build_alibi_tensor(max_seq_len, n_head, dtype=torch.bfloat16):
+    """
+    Alibi tensor is not causal as the original paper mentions, it relies on a translation invariance of softmax for
+    quick implementation: with l being a tensor, and a fixed value `softmax(l+a) = softmax(l) Based on
```
```suggestion
    quick implementation: with l being a tensor, and a fixed value `softmax(l+a) = softmax(l)`. Based on
```

sgugger reviewed (approved):
A few final nits and one comment marked as resolved which wasn't (please double-check you did fix the issues before resolving conversations :wink: ) and we should be good to merge.

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
Update docs/source/en/model_doc/bloom.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

younesbelkada committed:
minor changes:

- change docstring
- add link to paper

sgugger commented:
Good to merge once all tests are green (test hub failure can be ignored)!

patrickvonplaten commented:
@younesbelkada I think you can press the merge button now if you want ;-)

stas00 commented:
We need to send @younesbelkada a plaque with engraving:

`I authored a PR with 120 conversations and 195 commits and lived to tell the story!`

Amazing! 

I'd have given up long time ago.

thomasw21 commented:
Congratz!

younesbelkada commented:
The lights are green ! 
Thank you all!!
@stas00 if you count also the other PR there are 149 commits + 240 conversation

stas00 commented:
Yes, indeed! Good call, @younesbelkada!
